server/src/main/java/org/elasticsearch/cluster/coordination/DiscoveryUpgradeService.java

```diff
@@ -0,0 +1,321 @@
+                                    @Override
+                                    public void handleResponse(UnicastPingResponse response) {
+                                        long clusterStateVersion = UNKNOWN_VERSION;
+                                        for (PingResponse pingResponse : response.pingResponses) {
+                                            if (discoveryNode.equals(pingResponse.node())) {
+                                                clusterStateVersion
+                                                    = max(clusterStateVersion, pingResponse.getClusterStateVersion());
+                                            }
+                                        }
+                                        masterCandidates.add(new MasterCandidate(discoveryNode, clusterStateVersion));
+                                        listenableCountDown.countDown();
+                                    }
+
+                                    @Override
+                                    public void handleException(TransportException exp) {
+                                        logger.debug(
+                                            new ParameterizedMessage("unexpected exception when pinging {}", discoveryNode), exp);
+                                        listenableCountDown.onFailure(exp);
+                                    }
+
+                                    @Override
+                                    public String executor() {
+                                        return Names.SAME;
+                                    }
+
+                                    @Override
+                                    public UnicastPingResponse read(StreamInput in) throws IOException {
+                                        return new UnicastPingResponse(in);
+                                    }
+                                });
+
+                        } else {
+                            masterCandidates.add(
+                                new MasterCandidate(createDiscoveryNodeWithImpossiblyHighId(discoveryNode), UNKNOWN_VERSION));
+                            listenableCountDown.countDown();
+                        }
+                    }
+                    assert foundOldMaster;
+                }
+
+                @Override
+                public String toString() {
+                    return "discovery upgrade service retry";
+                }
+            });
+        }
+    }
+
+    /**
+     * Pre-7.0 nodes select the best master by comparing their IDs (as strings) and selecting the lowest one amongst those nodes with
+     * the best cluster state version. We want 7.0+ nodes to participate in these elections in a mixed cluster but never to win one, so
+     * we lie and claim to have an impossible ID that compares above all genuine IDs.
+     */
+    public static DiscoveryNode createDiscoveryNodeWithImpossiblyHighId(DiscoveryNode node) {
```

Comment:
can you add a unit test for this method, to make sure it indeed always compares greater to a randomly generated UUID.

Reply:
Yep, done in 212049c92fc09ae1e4a67428979bcc1bcddebe1e, see also the assertion in this method.